tests/unit/maxmemory.tcl

```diff
@@ -3,100 +3,130 @@ start_server {tags {"maxmemory"}} {
+        set evicted_clients [s evicted_clients]
+        set dbsize [r dbsize]
+        
+        puts "evicted clients: $evicted_clients, evicted keys: $evicted_keys, dbsize: $dbsize"
+        if $client_eviction {
+            assert {$evicted_clients > 0 && $evicted_keys == 0 && $dbsize == 50}
+        } else {
+            assert {$evicted_clients == 0 && $evicted_keys > 0 && $dbsize < 50}
+        }
     }
 
-    # TODO: we might want to change the tests so we run them first with no maxmemory-cliets and verify we get key eviction, and then re-run them and see clients disconnect and there's no eviction
-    r config set maxmemory-clients 3mb
-
-    set clients {}
-    test "eviction due to output buffers of many MGET clients" {
-        refill
+    foreach {client_eviction} {false true} {
+        set clients {}
+        test "eviction due to output buffers of many MGET clients, client eviction: $client_eviction" {
+            init_test $client_eviction
 
-        for {set j 0} {$j < 20} {incr j} {
-            set rr [redis_deferring_client]
-            lappend clients $rr
-        }
+            for {set j 0} {$j < 20} {incr j} {
+                set rr [redis_deferring_client]
+                lappend clients $rr
+            }
 
-        $rd debug sleep 1
-        $rd flush
-        after 100
+            $rd debug sleep 1
+            $rd flush
+            after 100
+            
+            for {set j 0} {$j < 5} {incr j} {
+                foreach rr $clients {
+                    $rr mget 1
+                    $rr flush
+                }
+            }
+            $rd read
 
-        for {set j 0} {$j < 5} {incr j} {
-            foreach rr $clients {
-                $rr mget 1
-                $rr flush
+            for {set j 0} {$j < 5} {incr j} {
+                foreach rr $clients {
+                    if {[catch { $rr read } err]} {
+                        lremove clients $rr
+                    }
+                }
             }
+
+            verify_test $client_eviction
+        }
+        foreach rr $clients {
+            $rr close
         }
-        $rd read
 
-        for {set j 0} {$j < 5} {incr j} {
+        set clients {}
+        test "eviction due to input buffer of a dead client, client eviction: $client_eviction" {
+            init_test $client_eviction
+            
+            for {set j 0} {$j < 30} {incr j} {
+                set rr [redis_deferring_client]
+                lappend clients $rr
+            }
+
             foreach rr $clients {
-                if {[catch { $rr read } err]} {
+                if {[catch {
+                    $rr write "*200\r\n"
+                    for {set j 0} {$j < 199} {incr j} {
+                        $rr write "\$1000\r\n"
+                        $rr write [string repeat x 1000]
+                        $rr write "\r\n"
+                        $rr flush
+                    }
+                }]} {
+                    puts "removing...."
```

Comment:
all `puts` must be inside a `$::verbose` condition.
normally, when the tests run in parallel by multiple runners, you don't want them to print anything to stdout on their own (instead they send messages to the parent to print on their behalf).

Reply:
my bad, forgot to remove it.